### ISSUE TITLE
Stairs rotate incorrectly via the Copy Paste Gadget

### ISSUE THREAD

**Issue by skitterywharf** [Direwolf20-MC/BuildingGadgets#478]
Forge version: 31.1.89
Building Gadgets version: 3.3.5

Hey, I installed the latest version of the mod into my 1.15.2 modpack and encountered an issue. Whenever I **Rotate** builds with the Copy Paste Gadget, the entire build rotates counter-clockwise, with the exception of stairs: they rotate clockwise.

![image](https://user-images.githubusercontent.com/32728173/82140968-19711800-984c-11ea-875c-bdb15976ea09.png)

As you can see, I built and copied a little compass to test this (blocks rotated north -> west; stairs rotated north -> east). I tested with Vanilla stairs and they do the same thing. 

I would like to copy builds from my Creative worlds into Survival and use Building Gadgets for that, any ideas what might be causing this?

**Comment by MajorTuvok** [comment]
I presume it is caused by some Rotation being inverted somewhere...

**Comment by MajorTuvok** [comment]
Sorry, but it really looks like I'm not going to do **anything** besides uni in this lecture period...
I hope that keeping this open reminds us/me that this is still a thing.

**Comment by MichaelHillcox** [comment]
I have a few ideas on why this might be but I won't be able to take a look at it for a few days

**Comment by MajorTuvok** [comment]
The relevant code is in the Template class...
There's a rotate Method somewhere and I presume it accidentally flips the rotation. If it were in the copy paste Gadget, then it would just rotate into the other direction.

**Comment by MichaelHillcox** [comment]
From what I can remember of the template system, the state data isn't being handle properly but I can take a look.

**Comment by MajorTuvok** [comment]
> the state data isn't being handle properly
?

**Comment by MichaelHillcox** [comment]
The rejection of state data isn't working, so I'd assume most features of it are some what broken, I'd have to look at it to remind me of what is actually wrong

**Comment by MajorTuvok** [comment]
> so I'd assume most features of it are some what broken
Rotation is completly unrelated to selecting irrelevant properties...
The point at which a state is rotated, it is already under the control of a Template instance and especially doesn't apply black lists. 
Applying a blacklist should happen either at creation time of a Template, or at build time of a Template, as this is not an inherent feature. It is rather something that we do in order to prevent some special data not getting into there.

**Comment by MichaelHillcox** [comment]
The second part does not happen as we've discussed the code that should do it was never hooked up thus why currently tile support is broken in many ways. 

I am unsure of the state of this code but during mine and dires time porting to 1.15 we found many issues with the copy-paste gadgets new systems.

**Comment by moomoo84** [comment]
We just need to pay attention to the ghosting outline before confirming the paste for now, If you face a certain way the blocks paste correctly from world to world. Looks like progress is being made, thank you dev's for taking the time to work through these issues , your doing a magnificent job. :)

**Comment by MajorTuvok** [comment]
> Looks like progress is being made,
Remember that we have very limited time... But thanks

**Comment by MichaelHillcox** [comment]
> We just need to pay attention to the ghosting outline before confirming the paste for now, If you face a certain way the blocks paste correctly from world to world. Looks like progress is being made, thank you dev's for taking the time to work through these issues , your doing a magnificent job. :) 

Yeah. I'll look into this soon as I'll be porting the mod in the coming days / week.

**Comment by hvdklauw** [comment]
Ok, just tested this on master branch, rotation seems to work there correctly, but placement seems to flipwhen rotating..

So the block on the left gets placed on the right and the other way around.

So a 90 degree rotation followed by a mirror places the blocks correctly.
180 degrees doesn't need a mirror
But 270 needs a mirror to be correct

**Comment by hvdklauw** [comment]
Ah, but that's only on a line, in 3 dimensions it's indeed all wrong...
but looking at the specific block transformation they seem to be correct, so maybe it's the rotationMatrixFor that's wrong?

**Comment by hvdklauw** [comment]
Found it.. the rotation matrix was indeed wrong !

https://github.com/Direwolf20-MC/BuildingGadgets/blob/master/src/main/java/com/direwolf20/buildinggadgets/common/util/tools/MathUtils.java#L120

The [2][0] needs to be sine and [0][2] needs to be -sine

**Comment by MichaelHillcox** [comment]
@MajorTuvok can you confirm so I can build out an update today or tomorrow to fix this and a few other outstanding issues?

**Comment by hvdklauw** [comment]
Had to look it up myself: https://en.wikipedia.org/wiki/Rotation_matrix#Basic_rotations

Seems this is also a bug in the 1.14 branch, want me to base my PR on 1.14?

**Comment by MichaelHillcox** [comment]
If you're working on a PR then feel free to make the same PR for both 1.14 and 1.15 as they are basically the same code bases

**Comment by hvdklauw** [comment]
Ok, I'll base it on 1.14, then it should be upmergable to 1.15 too :)
